app/controllers/application_controller.rb

```diff
@@ -61,15 +61,28 @@ def valid_help_actions
 
   def confirm_two_factor_setup
     user = request.env["warden"].user
-    return unless user &&
-      user.need_two_factor_authentication?(request) &&
-      !user.totp_enabled?
 
     if Rails.configuration.enable_enhanced_2fa_experience
+      return unless user
+
+      unless user.second_factor_method.nil?
+        # Handled by Devise's controller.
+        return if user.second_factor_method == "app"
+
+        # Handled by us.
+        redirect_to users_two_factor_authentication_email_path
```

Comment:
I'm not sure about this redirect here.  I believe this bit of code is supposed to check if the user has chosen email/app authentication. I did a bit of refactoring here which makes this a bit easier to read I think: https://github.com/alphagov/govwifi-admin/pull/1104/commits/7ed57ea221e7e41e5bbae3370a87164892bb2b3a

Reply:
Thanks Jos, my preference would be to land this as a separate PR. Let me know if you have any concerns with this.